packages/financial-templates-lib/src/logger/DiscordTicketTransport.ts

```diff
@@ -0,0 +1,94 @@
+// This transport enables winston logging to trigger opening of tickets through Discord bot. The transport configuration
+// requires bot authentication token and a dictionary of channel IDs. In order to send the ticket command this bot also
+// requires that the passed log object contains both message and mrkdwn strings, as well as discordTicketChannel
+// resolving to any of configured channel ID mapping.
+// This transport should be used with Ticket Tool (https://tickettool.xyz/) configured to trigger ticket commands on
+// the resolved channel ID and whitelisting of bot ID. Also the Discord server should be configured to allow the bot
+// to post messages on the ticket opening channel.
+import { Client, GatewayIntentBits } from "discord.js";
+import * as ss from "superstruct";
+import Transport from "winston-transport";
+
+import { removeAnchorTextFromLinks } from "./Formatters";
+import { isDictionary } from "./Logger";
+
+type TransportOptions = ConstructorParameters<typeof Transport>[0];
+
+const Config = ss.object({
+  botToken: ss.string(),
+  channelIds: ss.optional(ss.record(ss.string(), ss.string())),
+});
+// Config object becomes a type
+// {
+//   botToken: string;
+//   channelIds?: Record<string,string>;
+// }
+export type Config = ss.Infer<typeof Config>;
+
+// this turns an unknown ( like json parsed data) into a config, or throws an error
+export function createConfig(config: unknown): Config {
+  return ss.create(config, Config);
+}
+
+// Interface for log info object.
+interface DiscordTicketInfo {
+  message: string;
+  mrkdwn: string;
+  discordTicketChannel: string;
+}
+
+// Type guard for log info object.
+export const isDiscordTicketInfo = (info: unknown): info is DiscordTicketInfo => {
+  if (!isDictionary(info)) return false;
+  return (
+    typeof info.message === "string" && typeof info.mrkdwn === "string" && typeof info.discordTicketChannel === "string"
+  );
+};
+
+export class DiscordTicketTransport extends Transport {
+  private readonly botToken: string;
+  private readonly channelIds: { [key: string]: string };
+
+  private client: Client;
+
+  constructor(winstonOpts: TransportOptions, { botToken, channelIds = {} }: Config) {
+    super(winstonOpts);
+    this.botToken = botToken;
+    this.channelIds = channelIds;
+
+    this.client = new Client({ intents: [GatewayIntentBits.Guilds] });
+  }
+
+  // Note: info must be any because that's what the base class uses.
+  async log(info: any, callback: () => void): Promise<void> {
+    // We only try sending if we have all expected parameters and a matching channel ID.
+    const canSend = isDiscordTicketInfo(info) && info.discordTicketChannel in this.channelIds;
+
+    if (canSend) {
+      try {
+        if (!this.client.isReady()) await this.login(); // Log in if not yet established the connection.
```

Comment:
If multiple logs are triggered simultaneously, it may cause multiple logins. Should we do the login at the startup of the bot?

Reply:
Chatted with you about this and we decided that it's not a big problem and trying to do it differently would require bigger changes.

Reply:
Multiple calls to login worries me a bit... looking at the client implementation [here](https://github.com/discordjs/discord.js/blob/8f4256db8a52ac08359d0b3436f41b641ac4e382/packages/discord.js/src/client/Client.js#L213), it's unclear if it protects itself from concurrency issues.

Are we confident in this? If not, a fairly simple way to protect against these sorts of duplicate call/concurrency issues is:
```ts
export class DiscordTicketTransport extends Transport {
  private loginPromise: Promise<void> | null;

  ...

  async login(): Promise<void> {
    // Optional: handles cases where a log in expires or stops working for some reason
    if (!this.client.isReady()) {
      this.loginPromise = null;
    }
    // If the promise doesn't exist, set it.
    if (this.loginPromise === null) {
      this.loginPromise = this.client.login(this.botToken);
    }
    // Awaiting here will force everyone who calls concurrently to await the same promise instead of creating many promises.
    await this.loginPromise;
  }
}
```

Once again, totally fine with _not_ handling this if we're confident this sort of thing is already built in!